--- start-msvc10.bat.orig	2011-11-23 12:58:42 +0100
+++ start-msvc10.bat	2012-02-08 11:03:39 +0100
@@ -8,7 +8,7 @@
 echo "Mozilla tools directory: %MOZBUILDDIR%"
 
 REM Get MSVC paths
-call "%MOZBUILDDIR%guess-msvc.bat"
+call "%MOZBUILDDIR%guess-msvc-x64.bat"
 
 REM Use the "new" moztools-static
 set MOZ_TOOLS=%MOZBUILDDIR%moztools
@@ -85,4 +85,8 @@
 
 cd "%USERPROFILE%"
 
+if "%1"=="" (
 "%MOZILLABUILD%\msys\bin\bash" --login -i
+) else (
+"%MOZILLABUILD%\msys\bin\bash" --login -i -c "cd "%1" && make -f client.mk"
+)
