Index: src/vidalia/mainwindow.cpp
===================================================================
--- src/vidalia/mainwindow.cpp	(revision 2213)
+++ src/vidalia/mainwindow.cpp	(working copy)
@@ -260,6 +260,17 @@
     _torControl->stop();
   }
 
+  if (_browserProcess->state() != QProcess::NotRunning) {
+    /* The browser is still running -- ask it to close.
+       Currently this appears to have no affect, probably because killing
+       FirefoxPortable doesn't kill firefox.exe. Switching to kil() did
+       not help either. Once Vidalia starts firefox.exe directly this should work. */
+    // Ignore the finished signal
+    QObject::disconnect(_browserProcess, SIGNAL(finished(int, QProcess::ExitStatus)), 0, 0);
+    // Send the close request
+    _browserProcess->terminate();
+  }
+
   /* Disconnect all of the TorControl object's signals */
   QObject::disconnect(_torControl, 0, 0, 0);
 
